pydicom/dataset.py

```diff
@@ -581,14 +581,22 @@ def __delitem__(self, key):
         if isinstance(key, slice):
             for tag in self._slice_dataset(key.start, key.stop, key.step):
                 del self._dict[tag]
+                # invalidate private blocks in case a private creator is
+                # deleted - will be re-created on next access
+                if self._private_blocks and BaseTag(tag).is_private_creator:
+                    self._private_blocks = {}
```

Comment:
Perhaps my limited recollection of the private block code, but shouldn't this just delete the specific private block for that tag rather than clearing the whole dict?

Reply:
Yes, it could do that, but I found it easier (and especially faster) to clear it all, and have to do it only once. This is of course only relevant if you are deleting many tags (like with anonymization), but in this case you have to construct a base tag for the check only once. Also, if you are deleting many private tags, it is unlikely that you will access private tags afterwards, where you have to recreate the private blocks.
Anyway, that was my reasoning after the first try where I indeed tried to update only the affected block. Also, this way the code is much simpler... I can change that, of course.

Reply:
Took me a while to dig through this; I felt there might be a way to get into trouble.  I think I found one -- not a likely code sequence, but possible via our "API" of how private_block is expected to work:

``` python
from pydicom.dataset import Dataset

ds = Dataset()
block1 = ds.private_block(0x000b, 'dog^1', create=True)
block2 = ds.private_block(0x000b, 'dog^2', create=True)

block2b = ds.private_block(0x000b, 'dog^2') # works

del ds[0xb0010] # del block 1
block2b = ds.private_block(0x000b, 'dog^2') # errors
```
However, this is a more general problem (I think) that the code finding private blocks in the dataset stops on the first element number that doesn't exist yet.  So this might be a bigger discussion than here, and perhaps we did discuss it before when private blocks were introduced and I've just forgotten. 

Sorry, this is wordy and beyond the original intent of this PR, but I'd be interested in your thoughts.

Reply:
Yes, this actually came up in #1087 - what happens if there are holes in the private tag blocks. I searched the standard and didn't find anything that actually forbids that. There is this:
> The implementer shall insert an identification code in the first unused (unassigned) Element in this series to reserve a block of Private Elements.

but this is about writing tags - if we selectively delete private blocks, there may be holes. So to be correct, we shall seek for private creators until the last possible block, or we could compromise to a practical approach to not check more than some number (5? 10?) empty blocks, as in reality, more than 2-3 private creators in the same file are rare.
Searching all blocks doesn't take that long either, of course, depending on how often we need to do this.

I'm undecided, but tend to think that we shall support at least some cases with holes that are not completely improbable.

Reply:
BTW, this problem is independent of the current issue. Both ways to implement it - by specifically removing affected blocks, or by removing all blocks - will show this problem, as this is a problem in reading the blocks, not writing/deleting them.

Reply:
> BTW, this problem is independent of the current issue

Yes, I agree, the reading past holes is its own problem.

> Searching all blocks doesn't take that long either

I was wondering if dataset slicing can make this check easier.  In `Dataset.private_block()`, something like:
``` python

group_creators = {x.value:x for x in self[(group,0x10):(group,0x100)]}    
match = group_creators.get(private_creator)
```
then use the `match.tag.element` (if there is a match) to create the `PrivateBlock`.  Its equivalent to searching through in a loop, but seems tidier to me and universal, and quick enough if there aren't a large number of creators.

In any case, apologies for derailing this conversation. What do you think - deal with this now, or just come back to it after v2.0 release?

Reply:
> I was wondering if dataset slicing can make this check easier.

Yeah, that looks nicer. It probably doesn't make a difference performance-wise, but I alo like tidy.

> What do you think - deal with this now, or just come back to it after v2.0 release?

I think we shall deal with it now, as we are already fixing private block behavior. I may not do it today (bit tired), but certainly this week.

About the original issue: shall I change it to delete only the affected blocks?I would make it a bit more complicated, and deletion a bit slower, but reading an existing block the next time faster. What do you think?

Reply:
I think clearing `_private_blocks` as you have it now is fine if the reading part is fixed.  We could always come back to it, but I think the use cases which that would affect should be quite rare.  As you've said, the more likely case is removing all or many tags.



> I may not do it today (bit tired), but certainly this week.

Sure, no problem.

Reply:
Ok, I added something to support this - please have a look. The idea with slicing was actually a good one - I liked this!